lib/u2/ToggleListElementsView.es6.js

```diff
@@ -0,0 +1,234 @@
+  extends: 'foam.u2.Element',
+
+  documentation: `A view for toggling a list of arbitrary elements. Similar to
+      foam.u2.view.EditColumns view, but with a more consistent treatment of
+      list elements and the slots that bind them.
+
+      Elements in the selectable and selected lists are identified by their name
+      and use their label for rendering in UI. As such, list elements must each
+      have both a unique name and a label.`,
+
+  requires: [
+    'foam.u2.md.CheckBox'
+  ],
+  imports: ['error'],
+  exports: [
+    'selectable',
+    'selected',
+  ],
+
+  css: `
+    ^item {
+      display: flex;
+    }
+
+    ^ .foam-u2-md-CheckBox {
+      flex-shrink: 0;
+    }
+
+    ^ .foam-u2-md-CheckBox-label {
+      position: static;
+    }
+  `,
+
+  classes: [
+    {
+      name: 'ElementSlot',
+      extends: 'foam.core.Slot',
+
+      documentation: `A custom foam.core.Slot implementation; each manages one
+          element in the selected collection.`,
+
+      imports: [
+        'selectable',
+        'selected',
+      ],
+
+      properties: [
+        {
+          class: 'Int',
+          name: 'selectableIdx',
+        },
+      ],
+
+      methods: [
+        function get() {
+          const elem = this.selectable[this.selectableIdx];
+          return this.selected.some(c => c.name === elem.name);
+        },
+        function set(value) {
+          const prevValue = this.get();
+          if (value) {
+            this.addElement();
+          } else {
+            this.removeElement();
+          }
+          // Data binding depnds on an event being published on set. This is
+          // automatically taken care of by property slots (with the
+          // propertyChange event), but must be published manually for this
+          // custom foam.core.Slot implementation.
+          if (prevValue !== value) {
+            this.pub('elementChange', this.selectableIdx, this);
+          }
+        },
+        function getElement() {
+          return this.selectable[this.selectableIdx];
+        },
+        function addElement() {
+          const elem = this.selectable[this.selectableIdx];
+          const able = this.selectable;
+          const ed = this.selected;
+          let iAble = 0;
+          let iEd = 0;
+          while (iAble <= this.selectableIdx) {
+            if (iEd >= ed.length) {
+              ed.push(elem);
+              break;
+            }
+            if (iAble === this.selectableIdx) {
+              ed.splice(iEd, 0, elem);
+              break;
+            }
+
+            if (able[iAble].name === ed[iEd].name) iEd++;
+            iAble++;
+          }
+          this.selected = Array.from(ed);
+        },
+        function removeElement() {
+          const elem = this.selectable[this.selectableIdx];
+          const ed = this.selected;
+          for (let i = 0; i < ed.length; i++) {
+            if (ed[i].name === elem.name) {
+              ed.splice(i, 1);
+              i--;
```

Comment:
This method should only be able to remove one element, so you could break here, right?

And if so, could you not use find to get the index, and then feed that to splice?

Reply:
I guess I was going for remove all behaviour, but then the method name betrays the implementation. `removeElement`=remove first matching element found seems like a reasonable implementation.

Reply:
Done.